src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt

```diff
@@ -308,35 +308,53 @@ internal open class ParallelThreadsRunner(
-                    return ValidationFailureInvocationResult(scenario, validationResult.throwable)
+                    return ValidationFailureInvocationResult(scenario, validationResult.throwable, collectExecutionResults())
                 }
             }
             // Combine the results and convert them for the standard class loader (if they are of non-primitive types).
             // We do not want the transformed code to be reachable outside of the runner and strategy classes.
-            return CompletedInvocationResult(
-                ExecutionResult(
-                    initResults = initialPartExecution?.results?.toList().orEmpty(),
-                    parallelResultsWithClock = parallelPartExecutions.map { execution ->
-                        execution.results.zip(execution.clocks).map {
-                            ResultWithClock(it.first, HBClock(it.second.clone()))
-                        }
-                    },
-                    postResults = postPartExecution?.results?.toList().orEmpty(),
-                    afterInitStateRepresentation = afterInitStateRepresentation,
-                    afterParallelStateRepresentation = afterParallelStateRepresentation,
-                    afterPostStateRepresentation = afterPostStateRepresentation
-                )
-            )
+            return CompletedInvocationResult(collectExecutionResults(afterInitStateRepresentation, afterParallelStateRepresentation, afterPostStateRepresentation))
         } catch (e: TimeoutException) {
             val threadDump = collectThreadDump(this)
-            return RunnerTimeoutInvocationResult(threadDump)
+            return RunnerTimeoutInvocationResult(threadDump, collectExecutionResults())
         } catch (e: ExecutionException) {
-            return UnexpectedExceptionInvocationResult(e.cause!!)
+            return UnexpectedExceptionInvocationResult(e.cause!!, collectExecutionResults())
         } finally {
             resetState()
         }
     }
 
+    /**
+     * This method is called when we have some execution result other than [CompletedInvocationResult].
+     */
+    fun collectExecutionResults(): ExecutionResult {
+        return collectExecutionResults(null, null, null)
+    }
+
+    private fun collectExecutionResults(
+        afterInitStateRepresentation: String?,
+        afterParallelStateRepresentation: String?,
+        afterPostStateRepresentation: String?
+    ): ExecutionResult {
+        // We also collect results in case when some actors have not finished due to some reasons
+        // (execution has hung or unexpected exception).
+        // Therefore, some results may be null as corresponding actors won't finish.
+        // To process them correctly and pass only ready results, we do `takeWhile { it != null }` to cut off
+        // suffix with `null` results.
+        return ExecutionResult(
+            initResults = initialPartExecution?.results?.toList()?.takeWhile { it != null }.orEmpty(),
```

Comment:
Here you cut the `null` results, only to restore them later back in the reporter class. Let's avoid that, and just let `ExecutionResult` class contain `nulls` in the first place. This way we would also preserve the invariant that `scenarion` and `result` objects always have the same size.